src/tactic/scc.lean

```diff
@@ -12,66 +12,174 @@ of implications.
+meta def to_tactic_format (cl : closure) : tactic format :=
+do m ← read_ref cl,
+   let l := m.to_list,
+   fmt ← l.mmap $ λ ⟨x,y⟩, match y with
+                           | sum.inl y := pp y
+                           | sum.inr ⟨y,p⟩ := pformat!"({x}, {y}) : {infer_type p}"
+                           end,
+   pure $ to_fmt fmt
+
+meta instance : has_to_tactic_format closure := ⟨ to_tactic_format ⟩
+
+/-- `(n,r,p) ← root cl e` returns `r` the root of the tree that `e` is a part of (which might be itself)
+along with `p` a proof of `e ↔ r` and `n`, the preorder numbering of the root. -/
+meta def root (cl : closure) : expr → tactic (ℕ × expr × expr) | e :=
 do m ← read_ref cl,
    match m.find e with
    | none :=
      do p ← mk_app ``iff.refl [e],
-        pure (e,p)
-   | (some (e₀,p₀)) :=
-     do (e₁,p₁) ← root' e₀,
+        pure (0,e,p)
+   | (some (sum.inl n)) :=
+     do p ← mk_app ``iff.refl [e],
+        pure (n,e,p)
+   | (some (sum.inr (e₀,p₀))) :=
+     do (n,e₁,p₁) ← root e₀,
         p ← mk_app ``iff.trans [p₀,p₁],
-        modify_ref cl $ λ m, m.insert e (e₁,p),
-        pure (e₁,p)
+        modify_ref cl $ λ m, m.insert e (sum.inr (e₁,p)),
+        pure (n,e₁,p)
    end
 
-meta def root (cl : closure) (p : expr) : tactic (expr × expr) :=
-instantiate_mvars p >>= root' cl
+meta def merge_intl (cl : closure) (p e₀ p₀ e₁ p₁ : expr) : tactic unit :=
+do p₂ ← mk_app ``iff.symm [p₀],
+   p ← mk_app ``iff.trans [p₂,p],
+   p ← mk_app ``iff.trans [p,p₁],
+   modify_ref cl $ λ m, m.insert e₀ $ sum.inr (e₁,p)
 
+/-- `merge cl p`, with `p` a proof of `e₀ ↔ e₁` for some `e₀` and `e₁`, 
+merges the trees of `e₀` and `e₁` and keeps the root with the smallest preorder
+number as the root. This ensures that, in the depth-first traversal of the graph,
+when encountering an edge going into vertex whose equivalence class includes 
+a vertex that originated the current search, that vertex will be the root of 
+the corresponding tree. -/
 meta def merge (cl : closure) (p : expr) : tactic unit :=
 do `(%%e₀ ↔ %%e₁) ← infer_type p >>= instantiate_mvars,
-   (e₂,p₀) ← root cl e₀,
-   (e₃,p₁) ← root cl e₁,
+   (n₂,e₂,p₂) ← root cl e₀,
+   (n₃,e₃,p₃) ← root cl e₁,
    if e₂ ≠ e₃ then do
-     p₂ ← mk_app ``iff.symm [p₀],
-     p ← mk_app ``iff.trans [p₂,p],
-     p ← mk_app ``iff.trans [p,p₁],
-     modify_ref cl $ λ m, m.insert e₂ (e₃,p)
+     if n₂ < n₃ then do p ← mk_app ``iff.symm [p],
+                        cl.merge_intl p e₃ p₃ e₂ p₂
+                else cl.merge_intl p e₂ p₂ e₃ p₃
    else pure ()
 
+meta def assign_preorder (cl : closure) (e : expr) : tactic unit :=
+modify_ref cl $ λ m, m.insert e (sum.inl m.size)
+
+/-- `prove_eqv cl e₀ e₁` construct a proof of equivalence of `e₀` and `e₁` if
+they are equivalent -/
 meta def prove_eqv (cl : closure) (e₀ e₁ : expr) : tactic expr :=
-do (r,p₀) ← root cl e₀,
-   (r',p₁) ← root cl e₁,
+do (_,r,p₀) ← root cl e₀,
+   (_,r',p₁) ← root cl e₁,
    is_def_eq r r',
```

Comment:
I'm not sure, but would it be better to check for syntactic equality here? `r` and `r'` are props, right? It's unlikely, but I could imagine a situation where this starts unfolding much more than was intended. And it seems like we expect `r` and `r'` to be identical in the situations where they match here (but again I'm not certain).

Alternatively, maybe using `is_def_eq` with a low reducibility setting would be safer?

Reply:
You're right. I'm fixing it. Plus, if they are the same expressions, they will literally be represented by the same syntax trees at the same location in memory because it will the the root of both trees.